packages/runner/lib/workers/worker.ts

```diff
@@ -0,0 +1,153 @@
+            if (!fs.existsSync(url)) {
+                throw new Error(`RunnerWorker script not found at ${url.href}`);
+            }
+            this.worker = new Worker(url, { workerData: props });
+            this.worker.on('message', (msg: WorkerOutMessage) => {
+                if (msg.memoryInBytes) {
+                    this.memoryUsage = { memoryInBytes: msg.memoryInBytes, measuredAt: new Date() };
+                }
+            });
+        } else {
+            throw new Error('RunnerWorker should be instantiated in the main thread');
+        }
+    }
+
+    start(): void {
+        this.post('start');
+    }
+
+    abort(): void {
+        this.post('abort');
+    }
+
+    on(event: 'error', callback: (err: Error) => void): void;
+    on(event: 'exit', callback: (exitCode: number) => void): void;
+    on(event: 'error' | 'exit', callback: (arg: any) => void): void {
+        this.worker?.on(event, callback);
+    }
+
+    private post(message: WorkerInMessage): void {
+        this.worker?.postMessage(message);
+    }
+}
+
+export class RunnerChild {
+    private logger: ReturnType<typeof getLogger>;
+    private parent: MessagePort;
+    private jobsClient: JobsClient;
+    private heartbeatIntervalMs: number;
+    private memoryCheckIntervalMs: number;
+    private abortController: AbortController;
+    private taskId: string;
+    private execOpts: ExecOpts;
+
+    constructor({ parent, props }: { parent: MessagePort; props: WorkerProps }) {
+        if (isMainThread) {
+            throw new Error('RunnerChild should not be instantiated in the main thread');
+        }
+        this.jobsClient = new JobsClient({ baseUrl: props.jobsServiceUrl });
+        this.taskId = props.taskId;
+        this.heartbeatIntervalMs = props.heartbeatIntervalMs;
+        this.memoryCheckIntervalMs = props.memoryCheckIntervalMs;
+        this.abortController = new AbortController();
+        this.execOpts = {
+            nangoProps: props.nangoProps,
+            code: props.code,
+            codeParams: props.codeParams,
+            abortController: this.abortController,
+            locks: Locks.fromBuffer(props.locksBuffer)
+        };
+        this.logger = getLogger(`Worker-${this.taskId}`);
+
+        this.parent = parent;
+        this.parent.on('message', async (msg: WorkerInMessage) => {
+            switch (msg) {
+                case 'start':
+                    await this.start();
+                    break;
+                case 'abort':
+                    this.abort();
+                    break;
+            }
+        });
+    }
+
+    async run(): Promise<void> {
+        const heartbeat = setInterval(async () => {
+            try {
+                await this.jobsClient.postHeartbeat({ taskId: this.taskId });
+            } catch (err) {
+                this.logger.error('Heartbeat failed', { err });
+            }
+        }, this.heartbeatIntervalMs);
```

Comment:
should this be a setTimeout to avoid overlapping calls if jobs is unresponsive?

Reply:
good point